[Notifier] add RocketChat bridge

Jeroeny opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | See #33687
| License       | MIT

This would add [RocketChat](https://rocket.chat/) integration for the Notifier component. RocketChat is a self hosted chat service. 

Fully tested with a trial version of RocketChat.

@wirone I noticed you suggested this.

Jeroeny committed:
[Notifier] add RocketChat bridge

Wirone commented:
@Jeroeny yeah, I suggested it and even thought about implementation, but you were first 😉 👍

Jeroeny commented:
Is this good to go?

antiseptikk commented on src/Symfony/Component/Notifier/Bridge/RocketChat/RocketChatTransport.php:
```diff
@@ -0,0 +1,95 @@
+            throw new LogicException(sprintf('The "%s" transport only supports instances of "%s" for options.', __CLASS__, RocketChatOptions::class));
+        }
+
+        $options = $message->getOptions()->toArray();
```
Hi ! This line throw "Call to a member function toArray() on null" if you try to send message without options.

```
$message = new ChatMessage('Symfony TEST !');
```

Maybe you can test if `$message->getOptions()` is null ?

I tested with the following code but it may be possible to do better.

```
$options = $message->getOptions() instanceof MessageOptionsInterface ? $message->getOptions()->toArray() : [];
```

thePanz replied:
What about: `$options = ($opts = $message->getOptions()) ? $opts->toArray() : [];`
The same code is used in the telegram bridge

Jeroeny replied:
Thanks, I've changed it to that.

OskarStark commented on src/Symfony/Component/Notifier/Bridge/RocketChat/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.0.0
```
5.1.0

OskarStark commented on src/Symfony/Component/Notifier/Bridge/RocketChat/RocketChatOptions.php:
```diff
@@ -0,0 +1,60 @@
+/**
+ * @author Jeroen Spee <spee.jeroen@gmail.com>
+ *
+ * @experimental in 5.0
```
5.1

OskarStark commented on src/Symfony/Component/Notifier/Bridge/RocketChat/RocketChatTransport.php:
```diff
@@ -0,0 +1,95 @@
+ *
+ * @internal
+ *
+ * @experimental in 5.0
```
5.1

OskarStark commented on src/Symfony/Component/Notifier/Bridge/RocketChat/RocketChatTransportFactory.php:
```diff
@@ -0,0 +1,47 @@
+/**
+ * @author Jeroen Spee <spee.jeroen@gmail.com>
+ *
+ * @experimental in 5.0
```
5.1

OskarStark commented on src/Symfony/Component/Notifier/Bridge/RocketChat/composer.json:
```diff
@@ -0,0 +1,35 @@
+        }
+    ],
+    "require": {
+        "php": "^7.2.9",
```
7.2.5

OskarStark commented on src/Symfony/Component/Notifier/Bridge/RocketChat/composer.json:
```diff
@@ -0,0 +1,35 @@
+    "minimum-stability": "dev",
+    "extra": {
+        "branch-alias": {
+            "dev-master": "5.0-dev"
```
5.1-dev

Jeroeny commented:
Because I updated from the 5.0 upstream branch, which was wrong, I had 600+ commits as diff, also causing GitHub to request reviews from code owners. My bad, I reverted that push.

OskarStark reviewed (approved):
Looks good to me

fabpot commented on src/Symfony/Component/Notifier/Bridge/RocketChat/README.md:
```diff
@@ -0,0 +1,12 @@
+RocketChat Notifier
+==============
```
Missing `=`s

fabpot commented on src/Symfony/Component/Notifier/Bridge/RocketChat/RocketChatTransport.php:
```diff
@@ -0,0 +1,94 @@
+        string $chatChannel = null,
+        HttpClientInterface $client = null,
+        EventDispatcherInterface $dispatcher = null
+    ) {
```
Should be on one line

fabpot commented on src/Symfony/Component/Notifier/Bridge/RocketChat/RocketChatTransport.php:
```diff
@@ -0,0 +1,94 @@
+ */
+final class RocketChatTransport extends AbstractTransport
+{
+    protected const HOST = 'rocketchat.com';
```
Why is it protected and not private?

Jeroeny replied:
Inherited from the `AbstractTransport` parent. If it were private, it would throw an error: https://3v4l.org/sKOqV

fabpot commented on src/Symfony/Component/Notifier/Bridge/RocketChat/RocketChatTransportFactory.php:
```diff
@@ -0,0 +1,47 @@
+        if ('rocketchat' === $scheme) {
+            return (new RocketChatTransport($accessToken, $channel, $this->client, $this->dispatcher))
+                ->setHost($host)
+                ->setPort($port);
```
Should be on one line

fabpot commented:
@Jeroeny Can you rebase to get rid of the merge commit? Thank you.

fabpot commented:
Thank you @Jeroeny.
